### PR TITLE
Feat/deploy

### LINKED ISSUE #77: Implement deploy command
This needs to be completed: [lib/deploy.ts](https://github.com/NEARBuilders/bos-workspace/blob/main/lib/deploy.ts)

This should be able to work within a workspace or for a singular app (bos.workspace.json or just a bos.config.json).

You can reference the git workflow for deploying code in what the steps should be, see this: https://github.com/NEARBuilders/bos-workspace/blob/main/.github/workflows/deploy.yml

bw-legacy === false means use the current (v1.0.0-alpha.*) bos-workspace. We have this flag here because some projects are still using an old version of bos-workspace.

### PR DESCRIPTION
**Pull request by itechunter**
Draft #77

### CODE REVIEW

File: lib/cli.ts [hunk 1]
```diff
@@ -122,11 +123,21 @@ async function run() {
 
   program
     .command("deploy")
-    .description("Deploy the project (not implemented)")
-    .argument("[string]", "app name")
-    .action((appName) => {
-      console.log("not yet supported");
+    .description("Deploy the project")
+    .argument("[string]", "Workspace app name to deploy")
+    .option("--deploy-account-id <deployAccountId>", "Account under which component code should be deployed")
```
Comment (elliotBraem) [inline lib/cli.ts:128]: This all looks good, can you run `bw help` and update the README with this new command?

File: tests/unit/deploy.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      })
+    }) as any;
+    global.log = new Logger(LogLevel.ERROR);
+  })
+  afterAll(() => {
+    global.fetch = unmockedFetch;
+    global.log = unmockedLog;
+  })
+
+  it('should build correctly without logs', async () => {
```
Comment (elliotBraem) [inline tests/unit/deploy.ts]: I like how you've set up these tests so far -- for test cases, I'm thinking:

- [ ] should match expected input for bos-cli-rs (src/widget -> src), isolate that function in code call it "translate"
- [ ] should exec with correct command -> compare the the exec method call after running deployApp

File: lib/deploy.ts [hunk 3]
```diff
@@ -1,9 +1,102 @@
...
+};
 
 // deploy the app widgets and modules
-export async function deployAppCode(src: string, config: BaseConfig) {
+export async function deployAppCode(src: string, dist: string, opts: DeployOptions) {
+    const deploying = log.loading(`[${src}] Deploying app`, LogLevels.BUILD);
+
+    const config = await readConfig(path.join(src, "bos.config.json"), opts.network);
+
+    // Move files from "src/widget" to "src/"
```
Comment (elliotBraem) [inline lib/deploy.ts]: Looks great, how about moving L25-53 to a "translate" or "translateForBosCli" function so we can test it?

File: tests/unit/build.ts [hunk 4]
```diff
@@ -47,9 +47,9 @@ const app_example_1_output = {
   "/build/ipfs.json": JSON.stringify({
     "logo.svg": "QmHash",
   }, null, 2) + "\n",
-  "/build/src/widget/hello.utils.module.js": "const hello = (name) => `Hello, ${name}!`;\nreturn { hello };\n",
+  "/build/src/widget/hello/utils.module.js": "const hello = (name) => `Hello, ${name}!`;\nreturn { hello };\n",
```
Comment (elliotBraem) [inline tests/unit/build.ts]: It should be a dot, not /

**Comment by itechunter** [comment]
Resolve #77

File: tests/unit/build.ts [hunk 5]
```diff
@@ -47,9 +47,9 @@ const app_example_1_output = {
...
   "/build/src/widget/index.jsx": "const hello = \"hi\";\nreturn hello(props);\n",
-  "/build/src/widget/nested.index.jsx": "const hello = \"hi\";\nreturn hello(props);\n",
+  "/build/src/widget/nested/index.jsx": "const hello = \"hi\";\nreturn hello(props);\n",
```
Comment (elliotBraem) [inline tests/unit/build.ts]: Here too

File: tests/unit/build.ts [hunk 6]
```diff
@@ -118,7 +118,7 @@ const app_example_2_output = {
   "/build/ipfs.json": JSON.stringify({
     "logo.svg": "QmHash",
   }, null, 2) + "\n",
-  "/build/src/widget/hello.utils.module.js": "const hello = (name) => `Hello, ${name}!`;\nreturn { hello };\n",
+  "/build/src/widget/hello/utils.module.js": "const hello = (name) => `Hello, ${name}!`;\nreturn { hello };\n",
   "/build/src/widget/index.jsx": "const hello = \"hi\";\nreturn hello(props);\n",
```
Comment (elliotBraem) [inline tests/unit/build.ts]: Here

File: tests/unit/deploy.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    global.log = new Logger(LogLevel.ERROR);
+  })
+  afterAll(() => {
+    global.fetch = unmockedFetch;
+    global.log = unmockedLog;
+  })
+
+  it('should build correctly', async () => {
+    await deployAppCode('/app_example', '/build', {});
+    expect(vol.toJSON('/build')).toEqual(app_example_output);
```
Comment (elliotBraem) [inline tests/unit/deploy.ts:116]: This test is fine; can you please rename it so it's not conflicting with the build tests?

`should match expected input for bos-cli-rs`

**Comment by elliotBraem** [comment]
From testing it out locally

### deploy single

1. checkout branch, `yarn install`
2. `yarn build`
3. `cd examples/single`
4. `../../bin/bw.js deploy`

Although I did not provide any signing details, I still see a log for "App deployed successfully". I would expect a message that says "Necessary values not provided, please provide private & public key for deploy"

Screenshot 2024-04-27 at 11 29 33 PM

Otherwise works well

### deploy multi

3. `cd examples/multi`
4. `../../bin/bw.js deploy`

Although I am in a multi and I didn't provide an app name, it still attempts to deploy; it stalls on "deploying app"; we should stop it if in a workspace (has `bos.workspace.json`) and no app name provided

`../../bin/bw.js deploy HelloWorld` works well!

**Review by elliotBraem — APPROVED** [review]

**Pull request merged**